File: queue-api/src/main/java/net/mythicisland/queue/api/match/Match.java
package net.mythicisland.queue.api.match;

import net.mythicisland.queue.api.ticket.Ticket;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

/**
 * A set of tickets that will play together on one game server.
 *
 * @param id the unique ID of the match
 * @param queueType the queue type the match was created for
 * @param tickets the tickets forming the match
 * @param state the current state of the match
 * @param createdAt when the match was formed
 * @param assignment the allocated server, null while allocating
 * @param countdownEndsAt when the players get transferred, null until a server was allocated
 */
public record Match(
        UUID id,
        String queueType,
        List<Ticket> tickets,
        MatchState state,
        Instant createdAt,
        Assignment assignment,
        Instant countdownEndsAt
) {
}

File: queue-shared/src/main/kotlin/net/mythicisland/queue/shared/match/Match.kt
package net.mythicisland.queue.shared.match

import build.buf.gen.mythicisland.queue.v2.MatchState
import build.buf.gen.mythicisland.queue.v2.match
import net.mythicisland.common.util.protobuf.toTimestamp
import java.time.Instant
import java.util.UUID

data class Match(
    val id: UUID,
    val queueType: String,
    val ticketIds: List<UUID>,
    val state: MatchState,
    val createdAt: Instant,
    val assignment: Assignment? = null,
    val countdownEndsAt: Instant? = null,
) {

    fun toDefinition(tickets: List<Ticket>): build.buf.gen.mythicisland.queue.v2.Match {
        return match {
            id = this@Match.id.toString()
            queueType = this@Match.queueType
            this.tickets.addAll(tickets.map { it.toDefinition() })
            state = this@Match.state
            createdAt = this@Match.createdAt.toTimestamp()
            this@Match.assignment?.let { assignment = it.toDefinition() }
            this@Match.countdownEndsAt?.let { countdownEndTime = it.toTimestamp() }
        }
    }

}
